examples/nlp/mlm_and_finetune_with_bert.py

```diff
@@ -0,0 +1,531 @@
+    special_tokens=["[mask]"],
+)
+
+# Get mask token id for masked language model
+mask_token_id = vectorize_layer(["[mask]"]).numpy()[0][0]
+
+
+def encode(texts):
+    encoded_texts = vectorize_layer(texts)
+    return encoded_texts.numpy()
+
+
+def get_masked_input_and_labels(encoded_texts):
+    # 15% BERT masking
+    inp_mask = np.random.rand(*encoded_texts.shape) < 0.15
+    # Do not mask special tokens
+    inp_mask[encoded_texts <= 2] = False
+    # Set targets to -1 by default, it means ignore
+    labels = -1 * np.ones(encoded_texts.shape, dtype=int)
+    # Set labels for masked tokens
+    labels[inp_mask] = encoded_texts[inp_mask]
+
+    # Prepare input
+    encoded_texts_masked = np.copy(encoded_texts)
+    # Set input to [MASK] which is the last token for the 90% of tokens
+    # This means leaving 10% unchanged
+    inp_mask_2mask = inp_mask & (np.random.rand(*encoded_texts.shape) < 0.90)
+    encoded_texts_masked[
+        inp_mask_2mask
+    ] = mask_token_id  # mask token is the last in the dict
+
+    # Set 10% to a random token
+    inp_mask_2random = inp_mask_2mask & (np.random.rand(*encoded_texts.shape) < 1 / 9)
+    encoded_texts_masked[inp_mask_2random] = np.random.randint(
+        3, mask_token_id, inp_mask_2random.sum()
+    )
+
+    # Prepare sample_weights to pass to .fit() method
+    sample_weights = np.ones(labels.shape)
+    sample_weights[labels == -1] = 0
+
+    # y_labels would be same as encoded_texts i.e input tokens
+    y_labels = np.copy(encoded_texts)
+
+    return encoded_texts_masked, y_labels, sample_weights
+
+
+# We have 25000 examples for training
+x_train = encode(train_df.review.values)  # encode reviews with vectorizer
+y_train = train_df.sentiment.values
+train_classifier_ds = (
+    tf.data.Dataset.from_tensor_slices((x_train, y_train))
+    .shuffle(1000)
+    .batch(flags.BATCH_SIZE)
+)
+
+# We have 25000 examples for testing
+x_test = encode(test_df.review.values)
+y_test = test_df.sentiment.values
+test_classifier_ds = tf.data.Dataset.from_tensor_slices((x_test, y_test)).batch(
+    flags.BATCH_SIZE
+)
+
+# Build dataset for end to end model input (will be used at the end)
+test_raw_classifier_ds = tf.data.Dataset.from_tensor_slices(
+    (test_df.review.values, y_test)
+).batch(flags.BATCH_SIZE)
+
+# Prepare data for masked language model
+x_all_review = encode(all_data.review.values)
+x_masked_train, y_masked_labels, sample_weights = get_masked_input_and_labels(
+    x_all_review
+)
+
+mlm_ds = tf.data.Dataset.from_tensor_slices(
+    (x_masked_train, y_masked_labels, sample_weights)
+)
+mlm_ds = mlm_ds.shuffle(1000).batch(flags.BATCH_SIZE)
+
+"""
+## Create BERT model (Pretraining Model) for masked language modeling
+
+We will create a BERT-like pretraining model architecture
+using the `MultiHeadAttention` layer.
+It will take token ids as inputs (including masked tokens)
+and it will predict the correct ids for the masked input tokens.
+"""
+
+
+class BERTLayer(layers.Layer):
```

Comment:
You could specify this in 8-9 lines in the Functional API.

```
def bert_module(query, key, value):
   x = layers.MultiHeadAttention(...)(query, key, value)
   x = layers.Dropout(...)(x)
   x = layers.LayerNormalization(epsilon=1e-6)(x)
   x  = layers.Dense(...)(x)
   x  = layers.Dense(...)(x)
   x = layers.Dropout(...)(x)
   x = layers.LayerNormalization(epsilon=1e-6)(x)
   return x
```

Reply:
Agreed